### DISCUSSION TITLE
Oled output via i2c

### DISCUSSION THREAD

**Discussion by dmgeurts** [tjko/fanpico#6] [Ideas]
Would you see any issues with driving an i2c oled display for showing fan speeds or is the i2c bus exclusively used by the fanpico PCB?

**Comment by tjko** [comment]
Initial board design (with 4 inputs and 8 outputs) is using all available pins on Pico, so I2C bus is not available. 

But leaving I2C (and/or SPI) is a good idea, as it would allow all kinds of expansions to be easily added.
This would mean either reducing number of inputs or outpus (probably preferable to reduce input ports?), or adding some IC to multiplex some of the inputs to free up pins on Pico.

**Reply by dmgeurts** [reply]
I may, modify the board a little, as I don't need 8 outputs or 4 inputs for that matter.

I was more concerned about how to add the code for driving a display. My interest is showing fan tacho output and temperature readings.

**Reply by tjko** [reply]
Seems like there is couple common small OLED display module 'types':
- 0.96" (128x64 pixel)
- 0.91" (128x32 pixel)

Where you thinking something like these?  Or something larger?

I guess at least for these small ones should be able to add header on the PCB and possibly mounting points to easily add the OLED display?

**Reply by dmgeurts** [reply]
Yes, exactly that.

**Reply by dmgeurts** [reply]
My plan is to use the board in/on an eGPU case. I'd love to display two water temps and fan speeds in the display. Away from the main PCB for fan control. I only have 4 fans (280mm radiator) and a single input. That's not to say my use case trumps anyone else's, each implementation will have its own merits and challenges, but it may help to understand where I'm coming from.

Many years ago I was trained in electrical engineering and have used KiCAD since, so I'm happy to work on and submit an alternative board layout that would cater to a display (mounted or via header for wiring to somewhere else). My main struggle will be the required coding, I've used Pigpio for fan control and tacho reading with some success on a rPi 4, but a Pico is a different beast altogether.
 
Some years ago I ordered a small batch of PCBs here in the UK from a guy who collated orders, essentially splitting the overhead between a group of people, I'm sure I can find a way to get something produced again and made available in the UK.

**Reply by tjko** [reply]
I find the OLED display interesting option as well (would be nice to be able to just plug in a module on the PCB or use extension cable to mount it separately), I can certainly work on adding 'display' support to the firmware...

JLCPCB has certainly made prototyping easy, since they've brought the prices down so much that nowadays can just order 5-10 PCBs for under $10 (plus shipping).  I recall the current fanpico PCB costs under $10 for 5pcs and about $20 for 20 pieces when ordered from JLCPCB... (shipping is often more thant the PCB manufacturing cost...)

**Comment by tjko** [comment]
Freeing I2C should be relatively easy, just remove the MBFAN4 "input" connector and that would free up pins 24 and 25 on Pico.
Looks like these pins can be used as SDA1 and SCL1, so it should be possible to have board with I2C display support and up to 3 "motherboard" input and 8 fan connectors...

**Comment by tjko** [comment]
Preliminary support for (OLED) displays is now working...
[](url)
![fanpico-oled](https://user-images.githubusercontent.com/1075726/187586558-3c6f75f1-1b68-4e29-82d1-a810ae0f9ca2.jpeg)

**Comment by tjko** [comment]
It seems probably not worth it to bother with screens smaller than 128x64...

OLED display is now working with the basic information showing on the display:
![OLED status display](https://github.com/tjko/fanpico/blob/main/images/oled-status-screen-1.jpg?raw=true)
